graphql_compiler/schema_transformation/rename_schema.py

```diff
@@ -189,22 +192,27 @@ def __init__(self, renamings, query_type, scalar_types):
         self.builtin_types = frozenset({'String', 'Int', 'Float', 'Boolean', 'ID'})
 
     def _rename_name_and_add_to_record(self, node):
-        """Rename the value of the node, and add the name pair to reverse_name_map.
+        """Return a copy of input node with a new name, and add the name pair to reverse_name_map.
 
         Don't rename if the type is the query type, a scalar type, or a builtin type.
 
-        Modifies node and potentially modifies reverse_name_map.
+        The input node will not be modified. reverse_name_map may be modified.
 
         Args:
-            node: type Name (see graphql/language/ast), an object describing the name of an AST
-                  element such as type, interface, or scalar (user defined or builtin)
+            node: EnumTypeDefinition, InterfaceTypeDefinition, NamedTuple, ObjectTypeDefinition,
+                  or UnionTypeDefinition. An object representing an AST component, containing a
+                  .name attribute corresponding to an AST node of type Name.
+
+        Returns:
+            None, if the name was not changed; a Node object that is nearly identical to the
```

Comment:
The `visit` function works so that if `enter_(self, node, *args)` returns any value, even if the value is the exact same object `node` in the input, it will do shallow copies all the way up. It doesn't copy only when the return value is None. 

I could change it so that `_rename_name_and_add_to_record` always returns a Node object (returning the exact input if unchanged), and `enter` checks whether the output is the same object as the input, and if so, return `None`?

Reply:
Yeah, I think the change you proposed is reasonable.